Implement support for user-specified byte-ranged in S3's GetObject

binsenpai opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
When byte-ranges are used in web requests, sometimes we get formats such as `0-`, `-500`, or even `500-900`. The AWS S3 SDK for .NET has only supported having both the start of the range and the end of the range when requesting byte ranges. According to [RFC 7233 (Range Requests)](https://tools.ietf.org/html/rfc7233), the start value or end value of the range is not required and can be assumed by the server.

I've changed the `ByteRange.cs` to support a private value called `formattedByteRange`, which is a string that can be set by the user, in the public setter defined at `FormattedByteRange`. The getter of `FormattedByteRange` returns that value, or the formatted byte range based on the longs, if null. I've also added a second constructor which allows for setting the property directly as well. ~~`long`s that are `null` due to the fact that both can be null. When the getter for `FormattedByteRange` is called, it checks if both values are null, and if so, it returns a null `string`.~~ I've also modified `GetObjectRequest`'s `IsSetByteRange` to support checking whether both values are null as well. This prevents an empty header from being added to the request.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
I implemented this change because on one of my projects, I serve a file using the `ResponseStream` I get from the AWS SDK for .NET from my ASP.NET Core app. I also checked the request's headers for byte-ranges and would forward those to the AWS SDK for .NET, but the SDK doesn't support an empty starting or ending range, so I was stuck when Google Chrome tried to request ranges `0-`.

<!--- If it fixes an open [issue][issues], please link to the issue here -->
This change should provide a solution for, and fix the issue #886 

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I haven't done any proper testing yet, but my changes affect the way the SDK determines if a `byteRange` has been set or not, ~~and allowed for nullable `long`s in~~ added a constructor and a `formattedByteRange` value to `ByteRange.cs`, ~~which means that existing code *might* break. I'm confused about whether to set this as a breaking change or not, but I will.~~

I would appreciate if someone else added tests to this PR and test my code thoroughly, and if someone could place a PR for the documentation as well. Thanks.

## Screenshots (if appropriate)
None are available and none are needed.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
<!--- The SDK is released under the [Apache 2.0 license][license], so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a [Contributor License Agreement (CLA)][cla] -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

gokarnm commented:
@binsenpai , changing the properties from `long` to `long?` will be a binary breaking change. Given that Range header can have other values which cannot be set using just the Start and End properties, it would be better to add a setter to `FormattedByteRange` so you can directly provide a value for the range header. The `FormattedByteRange` getter can then use Start/End or FormattedByteRange based on which ones were set by user code.

Here are examples of range header that cannot be set using Start and End properties.

https://tools.ietf.org/html/rfc7233#page-5
```
   o  The first and last bytes only (bytes 0 and 9999):

        bytes=0-0,-1

   o  Other valid (but not canonical) specifications of the second 500
      bytes (byte offsets 500-999, inclusive):

        bytes=500-600,601-999
        bytes=500-700,601-999
```

binsenpai commented:
@gokarnm I understand, I'll go ahead and implement that in a few.

binsenpai commented:
@gokarnm Alright, I added the changes here. Feel free to have a look. Would it be all-right if I added a check to see if the "bytes=" value was present in the constructor and setter, and append it, if not?

binsenpai commented:
@normj Any chance we could get a review?

gokarnm commented:
@binsenpai , thanks for updating the PR! I'll review it and get back to you soon.

gokarnm commented on sdk/src/Services/S3/Custom/Model/ByteRange.cs:
```diff
@@ -61,7 +75,14 @@ public string FormattedByteRange
             get
             {
-                return string.Format(CultureInfo.InvariantCulture, "bytes={0}-{1}", this.Start, this.End);
+                if (this._formattedByteRange != null)
```
Check for empty string too, using string.IsNullOrEmpty.

gokarnm commented on sdk/src/Services/S3/Custom/Model/GetObjectRequest.cs:
```diff
@@ -156,7 +156,7 @@ public ByteRange ByteRange
         internal bool IsSetByteRange()
         {
-            return this.byteRange != null;
+            return this.byteRange != null && this.byteRange?.FormattedByteRange != null;
```
Nit, `byteRange?` seems unnecessary as null check is already done previously, I'd replace it with `this.byteRange.FormattedByteRange`.

gokarnm commented on sdk/src/Services/S3/S3.sln:
```diff
@@ -128,6 +128,46 @@ Global
 		{66F78F86-68D7-4538-8EA5-A669A08E1C19}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{66F78F86-68D7-4538-8EA5-A669A08E1C19}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{66F78F86-68D7-4538-8EA5-A669A08E1C19}.Release|Any CPU.Build.0 = Release|Any CPU
+		{78A15B4B-1533-4AA1-9E8C-571618A74FDB}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Could you exclude this file from the PR, it is autogenerated.

gokarnm commented:
@binsenpai , I've added some minor comments on the PR.

binsenpai commented:
@gokarnm Question, do you want me to exclude the file or exclude that specific line?

gokarnm commented:
Exclude the `sdk/src/Services/S3/S3.sln` file, as it's autogenerated.

binsenpai commented:
It's fine, I'll do it; I've got free time on my hands.

gokarnm commented:
@binsenpai , the PR has been [merged](https://github.com/aws/aws-sdk-net/commits?author=binsenpai).
